framework_agreement/model/portfolio.py

```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+
+        Boiler plate as OpenERP returns string instead of date/time objects...
+
+        :returns: namedtuple('AGDates', ['now', 'start', 'end'])
+        """
+        self.ensure_one()
+        AGDates = namedtuple('AGDates', ['now', 'start', 'end'])
+        now = fields.date.today()
+        start = datetime.strptime(self.start_date, DATE_FORMAT).date()
+        if self.end_date:
+            end = datetime.strptime(self.end_date, DATE_FORMAT).date()
+        else:
+            end = None
+        return AGDates(now, start, end)
+
+    def _search_state(self, operator, value):
+        """Search on the state field by evaluating on all records"""
+
+        all_records = self.search([])
+
+        if operator == '=':
+            found_ids = [a.id for a in all_records if a.state == value]
+        elif operator == 'in' and isinstance(value, list):
+            found_ids = [a.id for a in all_records if a.state in value]
+        elif operator in ("!=", "<>"):
+            found_ids = [a.id for a in all_records if a.state != value]
+        elif operator == 'not in' and isinstance(value, list):
+            found_ids = [a.id for a in all_records if a.state not in value]
+        else:
+            raise NotImplementedError(
+                'Search operator %s not implemented for value %s'
+                % (operator, value)
+            )
+        return [('id', 'in', found_ids)]
+
+    @api.multi
+    def create_new_agreement(self):
+        self.ensure_one()
+        new_agreement = self.env['product.pricelist'].create({
+            'name': '{}: New agreement'.format(self.name),
+            'portfolio_id': self.id,
+            'type': 'purchase',
+            'origin_address_id': self.supplier_id.id,
+            'version_id': [(0, 0, {
+                'name': '{}: Main version'.format(self.name),
+                'date_start': self.start_date,
+                'date_end': self.end_date,
+                'items_id': [(0, 0, {
+                    'name': '{}: Main rule'.format(self.name),
+                    'base': -2,
+                    'use_agreement_prices': True,
+                })],
+            })],
+        })
+
+        for product_line in self.line_ids:
+            self.env['product.supplierinfo'].create({
+                'name': self.supplier_id.id,
+                'agreement_pricelist_id': new_agreement.id,
+                'product_tmpl_id': product_line.product_id.product_tmpl_id.id,
+                'product_name': product_line.product_id.name,
+                'product_code': product_line.product_id.code,
+                'pricelist_ids': [(0, 0, {
+                    'min_quantity': 1.0,
+                    'price': 0.0,
+                })],
+            })
+
+    @api.multi
+    def get_line_for_product(self, product, quantity=0):
+        self.ensure_one()
+        for product_line in self.line_ids:
+            if (product_line.product_id == product and
```

Comment:
Hmm I guess we have an issue here. Nothing prevent to have many line with same product in a portfolio so the choice algo seems a bad fit.